### PR TITLE
Docs: document unittest class level setUp and tearDown compatibility

### PR DESCRIPTION
**Pull request by clebergnu**
This issue has come up repeatedly, so let's document how and why we
(don't) support `setUpClass` and `tearDownClass`.  While at it, let's
give some ideas on how to work around this, or better put, write tests
more aligned with Avocado's design of more isolated tests.

Signed-off-by: Cleber Rosa <crosa@redhat.com>

### CODE REVIEW

File: docs/source/WritingTests.rst [hunk 1]
```diff
@@ -992,6 +992,82 @@ Now, trying to list the tests on the ``mytest.py`` file again::
 You can also use the ``:avocado: disable`` tag, that works the opposite way:
 Something looks like an Avocado test, but we force it to not be listed as one.
 
+Unittest Compatibility Limitations And Caveats
+==============================================
+
+The Avocado test runner uses a different approach to running tests
+than most other Python unittest test runners.  This brings some
```
Comment (apahim) [inline docs/source/WritingTests.rst]: Too much white-spaces.

File: docs/source/WritingTests.rst [hunk 2]
```diff
@@ -992,6 +992,82 @@ Now, trying to list the tests on the ``mytest.py`` file again::
 You can also use the ``:avocado: disable`` tag, that works the opposite way:
 Something looks like an Avocado test, but we force it to not be listed as one.
 
+Unittest Compatibility Limitations And Caveats
+==============================================
+
+The Avocado test runner uses a different approach to running tests
```
Comment (apahim) [inline docs/source/WritingTests.rst]: This initial sentence is hard to get. Could you rephrase it?

File: docs/source/WritingTests.rst [hunk 3]
```diff
@@ -992,6 +992,82 @@ Now, trying to list the tests on the ``mytest.py`` file again::
 You can also use the ``:avocado: disable`` tag, that works the opposite way:
 Something looks like an Avocado test, but we force it to not be listed as one.
 
+Unittest Compatibility Limitations And Caveats
```
Comment (apahim) [inline docs/source/WritingTests.rst]: *Python* Unittest?

File: docs/source/WritingTests.rst [hunk 4]
```diff
@@ -992,6 +992,82 @@ Now, trying to list the tests on the ``mytest.py`` file again::
...
+==============================================
+
+The Avocado test runner uses a different approach to running tests
+than most other Python unittest test runners.  This brings some
+compatibility limitations that Avocado users should be aware.
+
+Execution Model
+---------------
+
+One of the main difference is a consequence of the Avocado design
```
Comment (apahim) [inline docs/source/WritingTests.rst]: "One of the main differences"

File: docs/source/WritingTests.rst [hunk 5]
```diff
@@ -992,6 +992,82 @@ Now, trying to list the tests on the ``mytest.py`` file again::
...
+decision that tests should be self contained and isolated from other
+tests.  Additionally, the Avocado test runner runs each test in a
```
Comment (apahim) [inline docs/source/WritingTests.rst]: Too much white-spaces.

File: docs/source/WritingTests.rst [hunk 6]
```diff
@@ -992,6 +992,82 @@ Now, trying to list the tests on the ``mytest.py`` file again::
...
+separate process.
+
+If you have a unittest class with many test methods and run them
+using most test runners, you'll find that all test methods run under
+the same process.  To check that behavior you could add to your
```
Comment (apahim) [inline docs/source/WritingTests.rst]: Too much white-spaces.

File: docs/source/WritingTests.rst [hunk 7]
```diff
@@ -992,6 +992,82 @@ Now, trying to list the tests on the ``mytest.py`` file again::
...
+If you run the same test under Avocado, you'll find that each test
+is run on a separate process.
+
+Class Level :meth:`setUp <unittest.TestCase.setUpClass>` and :meth:`tearDown <unittest.TestCase.tearDownClass>`
+---------------------------------------------------------------------------------------------------------------
+
+Because of Avocado's test execution model (each test is run on a
+separate process), it doesn't make sense to support unittest's
+:meth:`unittest.TestCase.setUpClass` and
+:meth:`unittest.TestCase.tearDownClass`.  Test classes are freshly
```
Comment (apahim) [inline docs/source/WritingTests.rst]: Too much white-spaces.

File: docs/source/WritingTests.rst [hunk 8]
```diff
@@ -992,6 +992,82 @@ Now, trying to list the tests on the ``mytest.py`` file again::
...
+instantiated for each test, so it's pointless to run code in those
+methods, since they're supposed to keep class state between tests.
+
+If you require a common setup to a number of tests, the current
+recommended approach is to to write regular :meth:`setUp
+<unittest.TestCase.setUp>` and :meth:`tearDown
+<unittest.TestCase.tearDown>` code that checks if a given state was
+set already set.  One example for such a test that requires a
```
Comment (apahim) [inline docs/source/WritingTests.rst]: Extra *set* and too much white-spaces.

File: docs/source/WritingTests.rst [hunk 9]
```diff
@@ -992,6 +992,82 @@ Now, trying to list the tests on the ``mytest.py`` file again::
...
+      def test(self):
+          process.run("%s 1" % self.sleep)
+
+If your test setup is some kind of action that will last accross
+processes, like the installation of a software package given in the
+previous example, you're pretty much covered here.
+
+If you need to keep other type of data a class acrross test
+executions, you'll have to resort to saving and restoring the data
+from an outside source (say a "pickle" file).  Finding and using a
```
Comment (apahim) [inline docs/source/WritingTests.rst]: Too much white-spaces.

File: docs/source/WritingTests.rst (hunk 2 above)
Comment (clebergnu) [inline docs/source/WritingTests.rst]: OK, how about:

"When executing tests, Avocado uses different techniques than most
other Python unittest runners.  This brings some compatibility
limitations that Avocado users should be aware."

File: docs/source/WritingTests.rst (hunk 1 above)
Comment (clebergnu) [inline docs/source/WritingTests.rst]: As discussed before, this is a style that I acquired earlier.  A good source of explanation on the reasons for using double-space is here:

http://creativepro.com/double-space-or-not-double-space/

IMHO, it helps when reading the sources, and it doesn't change anything when using the formatted version.  I'm not sure we should force a style for something that small at this point...

File: docs/source/WritingTests.rst (hunk 4 above)
Comment (clebergnu) [inline docs/source/WritingTests.rst]: OK!

File: docs/source/WritingTests.rst (hunk 8 above)
Comment (clebergnu) [inline docs/source/WritingTests.rst]: OK!

**Comment by clebergnu** [comment]
Closing in favor of #1512
